src/jobs/test.yml

```diff
@@ -48,27 +60,83 @@ steps:
       pkg-manager: <<parameters.pkg-manager>>
       cache-version: <<parameters.cache-version>>
       override-ci-command: <<parameters.override-ci-command>>
-  - when: # Run tests for NPM
+  - when: # Run tests for NPM, without test results
       condition:
-        equal: [npm, << parameters.pkg-manager >>]
+        and:
+          - equal: [npm, << parameters.pkg-manager >>]
+          - not: << parameters.test-results-for >>
       steps:
         - run:
             name: Run NPM Tests
             working_directory: <<parameters.app-dir>>
             command: npm run <<parameters.run-command>>
-  - when: # Run tests for YARN
+  - when: # Run tests for NPM, with jest test results
+      condition:
+        and:
+          - equal: [npm, << parameters.pkg-manager >>]
+          - equal: [jest, << parameters.test-results-for >>]
+      steps:
+        - run:
+            name: Check for test reporter
+            working_directory: <<parameters.app-dir>>
+            command: npm list jest-junit || (echo "Add the package jest-junit to the dev dependencies of your project with `npm install --save-dev jest-junit`" && exit 1)
+        - run:
+            name: Run NPM Tests
+            working_directory: <<parameters.app-dir>>
+            command: npm run <<parameters.run-command>> -- --reporters=default --reporters=jest-junit
+  - when: # Run tests for YARN, without test results
       condition:
-        equal: [yarn, << parameters.pkg-manager >>]
+        and:
+          - equal: [yarn, << parameters.pkg-manager >>]
+          - not: << parameters.test-results-for >>
       steps:
         - run:
             name: Run YARN Tests
             working_directory: <<parameters.app-dir>>
             command: yarn run <<parameters.run-command>>
-  - when: # Run tests for YARN 2.x
+  - when: # Run tests for YARN, with jest test results
       condition:
-        equal: [yarn-berry, << parameters.pkg-manager >>]
+        and:
+          - equal: [yarn, << parameters.pkg-manager >>]
+          - equal: [jest, << parameters.test-results-for >>]
+      steps:
+        - run:
+            name: Check for test reporter
+            working_directory: <<parameters.app-dir>>
+            command: yarn list --pattern jest-junit | grep " jest-junit@" || (echo "Add the package jest-junit to your projects dev dependencies with `yarn add --dev jest-junit`" && exit 1)
```

Comment:
Im sure you thought about this, what ultimately lead to deciding on an error, rather than automatically installing jest-junit at this point? There's definitely something to be said about not modifying a user's code automagically but I wonder if we might want to consider it.

Reply:
(1) It's really tricky to get installing something not in the dependencies to work in CI, because there are mechanisms in place to make sure builds are predictable (2) Someone who knows more about node dependency management than I do should consider the security implications.